


## Pan and Core genome

[Roary](https://sanger-pathogens.github.io/Roary/) the pan genome pipeline computes the number of orthologous genes in a number of core/pan spectrum partitions.

The core genome denotes the genes which are conserved between all samples (intersection), whereas the pan genome is the union of all genes across all samples.

```{r roary-table, echo=FALSE, message=F, warning=F, error=T}

roary_summary_file = paste0(results_directory, "/roary/summary_statistics.txt")


if (file.exists(roary_summary_file)) {

read_tsv(roary_summary_file, col_names = c("partition", "definition", "count")) %>% 

    custom_dt("roary", dom = "t", ordering = F, width = 500)

}

```

*Table `r tableno()`: Distribution of genes in different core/pan spectrum partitions.*



```{r roary-matrix, echo=FALSE, message=F, warning=F, error = T, fig.height = max(1.8, N*0.2), fig.width = 10}

plot_df = read_tsv(glob_list$roary) %>% 
    mutate(shared_by = rowSums(across(where(is.numeric)))) %>%
    pivot_longer(c(everything(), -Gene, -shared_by), names_to = "sample", values_to = "present") %>% 
    filter(present >= 1)  %>%

    # Two sorting options:
    arrange(shared_by, desc(sample)) %>% 
    #arrange(sample, shared_by) %>% 

    mutate(gene_order = 1:n()) %>% 
    mutate(gene_num = Gene %>% factor() %>% as.integer()) %>% 
    identity()



plot_df %>%
    ggplot(aes(reorder(Gene, gene_order), reorder(sample, desc(sample)), fill = factor(shared_by))) +
    geom_tile() +

    theme_classic() +
    theme(
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
    ) +
    labs(
        x = "gene",
        y = "sample",
        fill = "number of\nsamples",
    )

```

*Fig. `r figno()`: Genes shared between samples. Each vertical line represents a gene, and all lines have the same width regardless of the size of the gene. The genes are colored by the number of samples sharing them.*


---




