## `gtdb` - Genome taxonomy database taxonomical classification

[GTDB](https://gtdb.ecogenomic.org/) uses several public repositories with reference sequences and assigns the most likely name by measuring the average nucleotide identity (ANI) and relative evolutionary divergence (RED).


```{r gtdb, echo = F, message = F, warning = F, error = T}


gtdbtk_df = read_tsv(glob_list$gtdbtk, comment = "=", na = c("", "NA", "N/A")) %>% 
    mutate(user_genome_original = str_remove(user_genome, "^cm2_")) %>% 
    select(sample = user_genome_original, classification, fastani_ani, fastani_reference, msa_percent, red_value) %>%
    identity()


# Force the table to show missing samples by left joining onto the metadata table.
metadata_translate %>% 
    select(sample) %>% 
    left_join(gtdbtk_df) %>% 
    custom_dt("gtdbtk")
        


```

*Table `r tableno()`: Species classification provided by the GTDB-tk classify_wf workflow.*

---
