

## Other tools comparable to Assemblycomparator2

What is unique about Assemblycomparator2 is that it works strictly downstream of assembling and binning. Many other tools also include all the steps necessary to turn raw reads into genome representatives, and then does varying degrees of biological characterization of these freshly created assemblies/bins/genomes. It is a conscious decision to exclude the raw read-dependent tools out of the equation for Assemblycomparator2. This is because read-mapping, assembling or even binning is highly dependent on the sequencing technology used and requires a highly specialized pipeline for each technological use case. Assemblycomparator2 takes a different approach which is to offer a portable and flexible platform where you can easily compare your genomes, no matter where they came from, regardless of the sequencing technology used to create them in the first place. Genome quality is only increasing and in the future we will not have to be worried when comparing pyrosequencing and single-molecule sequencing or hybrid approach based genomes in a single batch of Assemblycomparator2. 

Below we are listing some competing pipelines that partly overlap with the use cases of Assemblycomparator2.
  - [Aviary](https://github.com/rhysnewell/aviary)
  - [Bactopia](https://github.com/bactopia/bactopia)
  - [DRAM](https://github.com/WrightonLabCSU/DRAM)
  - [Nullarbor](https://github.com/tseemann/nullarbor) 
  - [Tormes](https://github.com/nmquijada/tormes)
  - [VEBA](https://github.com/jolespin/veba)

